### ISSUE TITLE
purgeStale

### ISSUE THREAD

**Issue by ronag** [isaacs/node-lru-cache#213]
We have a live lock when calling `purgeStale` with 100% cpu.

v7.5.1

Sorry I don't have more information.

**Comment by deadbeef84** [comment]
I was debugging @ronag's issue, and it looked like it got stuck in `rindexes`

**Comment by isaacs** [comment]
What is the TTL on your cache?  What is the max?  How much time usually passes before it locks up?

**Comment by ronag** [comment]
```
hub$ npm ls lru-cache
hub@19.5.21 /Users/ronagy/GitHub/nxt/hub
├─┬ @nxtedition/lib@7.1.4
│ └── lru-cache@7.4.0
├─┬ @storybook/addon-docs@6.4.14
│ ├─┬ @storybook/builder-webpack4@6.4.14
│ │ ├─┬ terser-webpack-plugin@4.2.3
│ │ │ └─┬ cacache@15.3.0
│ │ │   └── lru-cache@6.0.0
│ │ └─┬ webpack@4.46.0
│ │   └─┬ terser-webpack-plugin@1.4.5
│ │     └─┬ cacache@12.0.4
│ │       └── lru-cache@5.1.1
│ └─┬ @storybook/core@6.4.14
│   └─┬ @storybook/core-server@6.4.14
│     └─┬ @storybook/manager-webpack4@6.4.14
│       ├─┬ terser-webpack-plugin@4.2.3
│       │ └─┬ cacache@15.3.0
│       │   └── lru-cache@6.0.0
│       └─┬ webpack@4.46.0
│         └─┬ terser-webpack-plugin@1.4.5
│           └─┬ cacache@12.0.4
│             └── lru-cache@5.1.1
├─┬ @storybook/builder-webpack5@6.4.14
│ └─┬ css-loader@5.2.7
│   └─┬ semver@7.3.5
│     └── lru-cache@6.0.0
├─┬ @storybook/manager-webpack5@6.4.14
│ └─┬ css-loader@5.2.7
│   └─┬ semver@7.3.5
│     └── lru-cache@6.0.0
├─┬ @storybook/react@6.4.14
│ └─┬ webpack@4.46.0
│   └─┬ terser-webpack-plugin@1.4.5
│     └─┬ cacache@12.0.4
│       └── lru-cache@5.1.1
├── lru-cache@7.5.1
└─┬ semver@7.3.5
  └── lru-cache@6.0.0
```

```
const activityCache = new LRU({ ttl: 10 * 60e3, max: 4096 })
```

> What are the types of keys that you're using?

Strings

> Does it lock up if you use 7.5.0?

Unable to test right now. It only happens in production so we have to revert to v6 for now. We'll see if we can get permission to try and repro.

**Comment by isaacs** [comment]
Are you sure it's 7.5.1 that's locking up, and not the 7.4.0 version being pulled in here?

```
├─┬ @nxtedition/lib@7.1.4
│ └── lru-cache@7.4.0
```

The purgeStale lockup was identified and fixed on v7.4.4.  There was still the possibility of an iteration busylock in general, however.  In 7.5.0, iteration busylock was possible, but purgeStale could not trigger it; it only happened only if _multiple_ list order mutations were made, or mutations were made to indexes other than the one being iterated over, so a single delete could not do it.  This was fixed more comprehensively in 7.5.1, so any invalid item encountered will abort the iteration.

**Comment by isaacs** [comment]
Hm, @nxtedition/lib doesn't use LRUCache with a ttl, though, so that can't be it.

**Comment by ronag** [comment]
@deadbeef84 I think we''re pretty sure it was 7.5.1.

Here is the full code we think triggers it:

```js

const activityCache = new LRU({ maxAge: 10 * 60e3 })
app.get('/activity/:activityId', (req, res, next) => {
  const { token } = req.cookies
  if (!token) {
    throw new createError.BadRequest('missing token')
  }

  const { id, username = id } = jwt.decode(token)
  if (!username) {
    throw new createError.BadRequest('missing username')
  }

  const { activityId } = req.params

  logger.debug({ activityId, username }, 'activity')

  activityCache.set(activityId, username)

  res.end()
})

Observable.timer(0, 10e3)
  .map(() => {
    const activity = {}

    activityCache.prune()
    activityCache.forEach((val, key) => {
      ;(activity[val] ??= []).push(key)
    })

    return {
      activity,
      activityCount: Object.keys(activity).length,
    }
  })
  .subscribe(stats$)
```

**Comment by isaacs** [comment]
Aha!  Got a repro, one second.  The key is that the list has to _not_ be full when purgeStale is being called, for some reason.

**Comment by isaacs** [comment]
Hm, scratch that, seems like I just had a bug in my test.  (This is why you should not call moveToTail without extreme care! lol)

**Comment by isaacs** [comment]
This test is passing reliably:

```js
  t.test('purgeStale() lockup exhaustive', t => {
    // range of ttls, from everything being stale to nothing being stale
    const ttls = [1, 10, 25, 300]
    const n = 10000
    const max = 10
    const fills = [0, 8, 9, 10, 11, 15]
    const shuffleLevels = [0, 0.1, 0.5, 0.9]
    for (const fill of fills) {
      for (const shuffle of shuffleLevels) {
        for (const ttl of ttls) {
          const name = `shuffle=${shuffle} fill=${fill} max=${max} ttl=${ttl}`
          t.test(name, t => {
            for (let i = 0; i < n; i++) {
              // create cache and fill it up
              const c = new LRU({
                max,
                ttl,
              })
              for (let j = 0; j < fill && j < max; j++) {
                clock.advance(3)
                c.set(j, j)
              }
              // now shuffle the list
              for (let j = 0; j < c.size; j++) {
                if (Math.random() < shuffle) {
                  c.moveToTail(j)
                }
              }
              // maybe add a bit more
              for (let j = max; j < fill; j++) {
                clock.advance(3)
                c.set(j, j)
              }
              c.purgeStale()
            }
            t.pass(`succeeded ${n} iterations`)
            t.end()
          })
        }
      }
    }
    t.end()
  })
```

I can't think of anything else that could be going on.  I can bring back the extra `toDelete` array in purgeStale, but without any more information about what's going on, to be able to reproduce and root cause it, it's just a guess, and I hate to give you something and say it's a fix when I'm not sure.  Down that path leads superstitious coding, never good.

Let me know if you figure out a way to reproduce it.

**Comment by ronag** [comment]
@isaacs Have you considered using fuzzing to test this lib?

**Comment by isaacs** [comment]
@ronag I often use fuzz tests (like the one in the comment above) to diagnose issues, but I prefer tests to be fast and deterministic, so once the issue is reproduced, I root cause it and create a more targeted unit test that just creates the bad state deliberately and as minimally as possible.  Having a fuzz test fail randomly in CI is not useful if it can't be reproduced locally.

**Comment by deadbeef84** [comment]
@ronag this actually happened in hub v19.5.19, which has lru-cache `7.4.2`. I double checked because I remember rindex function was looking different when I debugged compared to current implementation (7.5.1). Case closed?

**Comment by deadbeef84** [comment]
I'm actually able to reproduce the issue using the values from the log if you're interested

**Comment by isaacs** [comment]
@deadbeef84 If you have data to reproduce the issue using lru-cache 7.5.1, then I'm definitely interested.  The specific keys/values don't matter, what's important is the state of the cache.  Specifically: the next/prev index arrays, and the ttls/starts lists.  But I can calculate that if you have the sequence of time stamps of set/get calls and the time when `purgeStale` was called.

**Comment by deadbeef84** [comment]
Ah sorry, I should have been more clear, I can only reproduce the issue on 7.4.2, not 7.5.1.

**Comment by isaacs** [comment]
Well, yes, 7.4.2 has the issue, it was fixed in 7.4.4.

70a0b04b03d6404bbcb7f97ab8a8b7d04be71a84 adds a test that calls purge stale on every possible ordering and staleness of 5 items in the cache (3840 possible combinations).  Closing this unless someone can show a reproduction case, I'm confident that 7.5.1 does not exhibit a busylock on purgeStale.  The simplest explanation is that it was a different version of lru-cache 7 which did have this issue.
